ssl/ssl_cert_comp.c

```diff
@@ -0,0 +1,484 @@
+                    continue;
+
+                if (!ssl_compress_one_cert(ssl, &cpks[j], alg))
+                    return 0;
+
+                /* if the cert expanded, set the value in the CERT_PKEY to NULL */
+                if (cpks[j].comp_cert[alg]->len >= cpks[j].comp_cert[alg]->orig_len) {
+                    OSSL_COMP_CERT_free(cpks[j].comp_cert[alg]);
+                    cpks[j].comp_cert[alg] = NULL;
+                } else {
+                    count++;
+                }
+            }
+        }
+    }
+    return (count > 0);
+}
+
+static size_t ssl_get_compressed_cert(SSL *ssl, CERT_PKEY *cpk, int alg, unsigned char **data, size_t *orig_len)
+{
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(ssl);
+    size_t cert_len = 0;
+    size_t comp_len = 0;
+    unsigned char *cert_data = NULL;
+    OSSL_COMP_CERT *comp_cert = NULL;
+
+    if (sc == NULL
+            || cpk == NULL
+            || data == NULL
+            || orig_len == NULL
+            || !sc->server
+            || !SSL_in_before(ssl))
+        return 0;
+
+    if ((cert_len = ssl_get_cert_to_compress(ssl, cpk, &cert_data)) == 0)
+        goto err;
+
+    comp_cert = OSSL_COMP_CERT_uncompressed_data(cert_data, cert_len, alg);
+    OPENSSL_free(cert_data);
+    if (comp_cert == NULL)
+        goto err;
+
+    comp_len = comp_cert->len;
+    *orig_len = comp_cert->orig_len;
+    *data = comp_cert->data;
+    comp_cert->data = NULL;
+ err:
+    OSSL_COMP_CERT_free(comp_cert);
+    return comp_len;
+}
+
+static int ossl_set1_compressed_cert(CERT *cert, int algorithm,
+                                     unsigned char *comp_data, size_t comp_length,
+                                     size_t orig_length)
+{
+    OSSL_COMP_CERT *comp_cert;
+
+    /* No explicit cert set */
+    if (cert == NULL || cert->key == NULL)
+        return 0;
+
+    comp_cert = OSSL_COMP_CERT_compressed_data(comp_data, comp_length, orig_length, algorithm);
+    if (comp_cert == NULL)
+        return 0;
+
+    OSSL_COMP_CERT_free(cert->key->comp_cert[algorithm]);
+    cert->key->comp_cert[algorithm] = comp_cert;
+
+    return 1;
+}
+#endif
+
+/*-
+ * Public API
+ */
+int SSL_CTX_set1_cert_comp_preference(SSL_CTX *ctx, int *algs, size_t len)
+{
+#ifndef OPENSSL_NO_COMP_ALG
+    return ssl_set_cert_comp_pref(ctx->cert_comp_prefs, algs, len);
+#else
+    return 0;
+#endif
+}
+
+int SSL_set1_cert_comp_preference(SSL *ssl, int *algs, size_t len)
+{
+#ifndef OPENSSL_NO_COMP_ALG
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(ssl);
+
+    if (sc == NULL)
+        return 0;
+    return ssl_set_cert_comp_pref(sc->cert_comp_prefs, algs, len);
+#else
+    return 0;
+#endif
+}
+
+int SSL_compress_certs(SSL *ssl, int alg)
+{
+#ifndef OPENSSL_NO_COMP_ALG
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(ssl);
+    CERT_PKEY *cpks;
+
+    if (sc == NULL)
+        return 0;
+    if (sc->cert != NULL)
+        cpks = sc->cert->pkeys;
+    else
+        cpks = ssl->ctx->cert->pkeys;
```

Comment:
This looks wrong. `ssl_compress_certs` modifies the underlying `CERT_PKEY` structure. We can't do that on the parent  SSL_CTX since it might not be thread safe.